core/arch/interp.c

```diff
@@ -5771,6 +5771,183 @@ mangle_x64_ib_in_trace(dcontext_t *dcontext, instrlist_t *trace, instr_t *target
 }
 #endif
 
+#ifdef AARCH64
+/* Prior to indirect branch trace mangling, we check previous byte blocks
+ * before each indirect branch if they were mangled properly.
+ * An indirect branch:
+ *    br jump_target_reg
+ * is mangled into (see mangle_indirect_jump in mangle.c):
+ *    str IBL_TARGET_REG, TLS_REG2_SLOT
+ *    mov IBL_TARGET_REG, jump_target_reg
+ *    b ibl_routine or indirect_stub
+ * This function is used by mangle_indirect_branch_in_trace;
+ * it removes the two mangled instructions
+ * and returns the jump_target_reg id.
+ */
+#    define STR_IMM_ENC 0xf9000000
+#    define SAVE_IBL_REG_ENC                                                       \
+        ((STR_IMM_ENC | 16 >> 3 << 10) | (dr_reg_stolen - DR_REG_START_GPR) << 5 | \
+         (IBL_TARGET_REG - DR_REG_START_GPR))
+#    define MOV_IBL_REG_ENC_MASK 0xff00ffff
+#    define MOV_IBL_REG_MASK 0x001f0000
+#    define MOV_IBL_REG_ENC (0xaa0003e0 | (IBL_TARGET_REG - DR_REG_START_GPR))
+
+static reg_id_t
+check_and_remove_patched_ibl(instrlist_t *trace, instr_t *targeter)
+{
+    int i, j;
+    instr_t *prev = instr_get_prev(targeter);
+    /* XXX: Currently we assume the previous instruction is undecoded. */
+    if (!instr_raw_bits_valid(prev)) {
+        ASSERT_CURIOSITY(false);
+        return DR_REG_NULL;
+    }
+    uint num_instr = prev->length / AARCH64_INSTR_SIZE;
```

Comment:
I may be missing something here, but isn't prev->length always `AARCH64_INSTR_SIZE` (unless its a label or ldstex)?

Reply:
A L0 instruction can contain multiple instructions, I believe?